designdocs/provider_registration.md

```diff
@@ -0,0 +1,189 @@
+1. User stakes balance to become a provider. This is called a "Registration Fee".
+1. The "Registration Fee" is credited to the Chain Treasury (TBD).
+1. The provider's MSA id is added to storage on chain.
+
+Should a provider choose at any point to stop providing services, it can issue a
+request to remove itself from the provider registry:
+1. The provider issues an announcement that it will stop providing services.
+1. The provider's MSA id is removed from storage.
+
+**Is the provider's Registration Fee reimbursed at this point?**
+
+Please note:
+* All names are placeholders and may be changed.
+* Types may change as needed during implementation phase
+* Errors in the extrinsic(s) must have different, reasonably-named error enums
+  for each type of error for ease of debugging.
+
+### Types
+* `ProviderRegistrationParams<T: Config>`, the arguments used to register an announcement.
+  * `provider_msa_id`: `MsaId`
+  * `registration_fee`: `Balance` **Question: will the fee be the same for all providers? If so, we may not need to include it in the announcement**
+  * `provider_metadata`: `ProviderAnnouncementMetadata`
+* `ProviderRegistrationAnnouncement<T: Config>`, the resource that exists on-chain
+  * `block_number`: `BlockNumber`
+  * `provider_msa_id`: `MsaId`
+  * `provider_metadata`: `ProviderAnnouunucementMetadata`
+* `ProviderUnregistrationParams<T:Config>`, the arguments used to unregister an announcement.
+  * `provider_msa_id`: `MsaId`
+* `ProviderUnregistrationAnnouncement<T:Config>`, the resource that exists on-chain
+  * `provider_msa_id`: `MsaId`
+* `ProviderAnnouncementMetadata`
+  * `name`: `Vec<u8>`
+* `Provider`
+  * `provider_msa_id`: `MsaId`
+  * `name`: `Vec<u8>`
+* `ProviderRegistrationAnnouncementOptions<T:Config>`
+  * `provider_msa_id`:  `Option<MsaId>`, the announcer's MSA id.  Pass None() to get all announcements.
+  * `from_block`: `<T::BlockNumber>`, retrieve messages starting at the given block number (inclusive)
+  * `to_block`: `<T::BlockNumber>`, retrieve messages ending at the given block number (inclusive)
+  * `from_index`: `u32`, starting message index
+  * `page_size`: `usize`, retrieve `page_size` messages at a time, up to configured `T::PageSizeMax`. If 0, return `T::PageSizeMax` results
+
+#### Storage
+* `ProviderRegistry<T: Config>`: `StorageMap<MsaId, Provider>`
+  * Stores registered providers and provides lookup functionality via `MsaId`.
+    Existence in this storage structure implies that a provider's fee has been
+    paid and their registration was otherwise successful.
+* `ProviderRegistrationFees<T: Config>`: `StorageMap<MsaId, Balance>`
+  * Stores registration fees paid by providers. This structure can be used to
+    verify that a provider has indeed paid their registration fee.
+    **NOTE: This may not be necessary. It is being included here to allow for fee verification in the case where we have limited insight into the chain treasury.**
```

Comment:
This may not be necessary. It is being included here to allow for fee verification in the case where we have limited insight into the chain treasury.